File: web/src/krate/chnot/dto.ts
import type { TID } from "@/lib/id_util";
import type { DbText, Varchar } from "@/lib/types";
import type { MdwtTagSearchType } from "../mdwt/dto";
import type { TodoEvent } from "../toent/toent-model";
import type { ChnotKind, ChnotMeta } from "./po";

export type ChnotThreadOrderCommitReqData = {
  otid: TID;
  closed: boolean;
};
export type ChnotSearchReq = {
  query?: string;
  tags?: MdwtTagSearchType;
  kinds: ChnotKind[];
  start_index: number;
  page_size: number;
};
export type ChnotThreadMetaFetchReq = {
  otid: TID;
  include_hist?: boolean;
};
export type ChnotThreadMetaFetchRsp = {
  chnot_meta_sorted: ChnotThreadMetaFetchRspData[];
};

export type ChnotThreadOrderCommitReq = {
  thread_otid: TID;
  orders: ChnotThreadOrderCommitReqData[];
  remove_others?: boolean;
};
export type ChnotThreadOrderCommitRsp = object;

export type ChnotMetaCommitReqData = {
  otid: TID;
  kind: ChnotKind;
  kspace: Varchar<40>;
  archive?: boolean;
  pin_it?: boolean;
};
export type ChnotMetaCommitReq = {
  metas: ChnotMetaCommitReqData[];
};
export type ChnotMetaCommitRsp = {
  metas: ChnotMeta[];
};

export type ChnotMetaListReq = {
  otids: TID[];
};
export type ChnotMetaListRsp = {
  metas: ChnotMeta[];
};

export type ChnotThreadMetaFetchRspData = {
  meta: ChnotMeta;
  closed: boolean;
  heading_level?: number;
};

export type ChnotSearchRspData = {
  title?: string;
  meta: ChnotMeta;
};

export type ChnotThreadOrderArchiveRsp = object;
export type ChnotThreadOrderArchiveReq = {
  thread_otid: TID;
  otids: TID[];
};

File: lib/backend/src/krate/chnot/dto.rs
use chin_sql::{str_type::Varchar, time_type::TID};
use serde::{Deserialize, Serialize};

use crate::krate::mdwt::MdwtTagSearchType;

use super::*;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotSearchRspData {
    pub title: Option<String>,
    pub meta: ChnotMeta,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadOrderCommitReqData {
    pub otid: TID,
    pub closed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadOrderCommitReq {
    pub thread_otid: TID,
    pub orders: Vec<ChnotThreadOrderCommitReqData>,
    pub remove_others: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadOrderCommitRsp {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotMetaCommitReqData {
    pub otid: TID,
    pub kind: ChnotKind,
    pub kspace: Varchar<40>,
    pub archive: Option<bool>,
    pub pin_it: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotMetaCommitReq {
    pub metas: Vec<ChnotMetaCommitReqData>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotMetaCommitRsp {
    pub metas: Vec<ChnotMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotMetaListReq {
    pub otids: Vec<TID>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotMetaListRsp {
    pub metas: Vec<ChnotMeta>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotSearchReq {
    pub query: Option<String>,

    pub tags: Option<MdwtTagSearchType>,
    pub kinds: Vec<ChnotKind>,

    with_archive: Option<bool>,
    hide_thread: Option<bool>,

    // Paging
    pub start_index: usize,
    pub page_size: usize,
}

impl ChnotSearchReq {
    pub fn with_archive(&self) -> bool {
        self.with_archive.unwrap_or(false)
    }

    pub fn hide_thread(&self) -> bool {
        self.hide_thread.unwrap_or(true)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadMetaFetchReq {
    pub otid: TID,
    pub include_hist: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadMetaFetchRspData {
    pub meta: ChnotMeta,
    pub closed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadMetaFetchRsp {
    pub chnot_meta_sorted: Vec<ChnotThreadMetaFetchRspData>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadOrderArchiveRsp {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChnotThreadOrderArchiveReq {
    pub thread_otid: TID,
    pub otids: Vec<TID>,
}
